### ISSUE TITLE
Add pop-up or other obvious notification for getting a message or changeset comment

### ISSUE THREAD

**Issue by aawiseman** [openstreetmap/openstreetmap-website#3182]
I've seen that mappers don't often realize they have gotten a message or a changeset comment, which can be a problem in general, but especially if there are new mappers or people who are not following policies and may not realize it. Even if you tell new mappers that they might receive messages or changeset comments, it's not very clear or obvious that they have received one.

I think it would be good to have a pop-up window, blinking notification or some other obvious visual showing that they just received a message or changeset comment -- ideally as soon as it arrives if they are are looking at osm.org, and as soon as they log in if they aren't at osm.org. 

I think this could help with new mapper retention as well, since there have been studies like this that people who receive a comment or message are more likely to keep mapping https://core.ac.uk/download/pdf/159755161.pdf

**Comment by tomhughes** [comment]
Well we send an email already, which will arrive even if they don't visit the site!

**Comment by aawiseman** [comment]
Right, I just think an obvious visual on the website would help as well -- a pop-up they can't ignore, some kind of blinking or flashing thing, etc. They might not think to look for the email, it might go to spam, etc. 

I think in general the more ways people can see they are part of a community and are getting messages/comments, the better. Especially for new mapppers, an email saying "has commented on one of your changesets" is not super clear. 

Or maybe they only planned to map one time, so they might not even notice or care if they get an email, but if they get a notification while they're mapping, it might help to encourage them to keep mapping.

**Comment by migurski** [comment]
Andrew, do you mean the same information as the unread-count next to My Messages in the account drop-down?

<img width="741" alt="Screen Shot 2021-04-23 at 4 07 03 PM" src="https://user-images.githubusercontent.com/58730/115937924-f41e5e00-a44d-11eb-9194-29fd5acfad51.png">

LinkedIn does a reasonable job of showing counts attached to top nav, we could imitate this.

<img width="531" alt="Screen Shot 2021-04-23 at 4 07 20 PM" src="https://user-images.githubusercontent.com/58730/115937941-fda7c600-a44d-11eb-80a5-9dd67b34100d.png">

**Comment by aawiseman** [comment]
I think something more obvious than that -- for example I've seen new mappers not even notice that little number while looking over their shoulder and teaching them to map, and when I say "hey you got a message" it's not obvious where to look.

I think an actual popup on the screen would be ideal -- "You just received a message! Click here to read" or "You just received a comment on one of your edits! Click here to read", which then goes to their message/comment. Or at least some kind of blinking or flashing thing on the top menu bar that makes it more obvious.

**Comment by SomeoneElseOSM** [comment]
(somewhat related) - a list of "all the messages that you have received" would be useful too.  That exists, at Pascal Neis' site (see e.g. http://resultmaps.neis-one.org/osm-discussion-comments?uid=61942 ) but it's not straightforward to navigate to from osm.org (select a changeset, look at xml select uid from xml, browser to the link in the above format with that uid at the end).

**Comment by tomhughes** [comment]
There is no concept of having "read" a comment (other than for messages) at the moment, and it would be a considerable challenge because you would have to track the read state for each comment for every subscriber to a changeset.

A modal popup is also an extremely "in your face" approach. I understand that may be your intention, but it probably wouldn't be welomed by all.

**Comment by tomhughes** [comment]
@SomeoneElseOSM well that's the thing - the object model doesn't consider them as personal messages to an individual, they are public comments on something that you happen to be subscribed to.

**Comment by migurski** [comment]
Another possibility here: often apps will differentiate between the first time something happens and subsequent times. I don’t have an example to hand, but the popup or notification might be used just once per account, to familiarize a mapper with the idea that they have inbound messages but then not nagging them again later.

**Comment by tomhughes** [comment]
In any case this is all basically already covered in things like #1387 and #2377.

**Comment by migurski** [comment]
I agree that those comments from almost a half-decade ago concern a similar topic. 😆

**Comment by aawiseman** [comment]
That's a good point, maybe the pop-up happens with just the first message and first changeset comment someone gets. 

At bare minimum I think at least making the little "unread" notification on the upper right corner more obvious would be a big improvement I think, similar to @migurski 's comment above with the Linkedin example. Making it red or blink or flash or something that signals "hey you, look here!" 

Of course, that only covers direct messages, it would also be good to do something to cover changeset comments. I bet those are more comment than direct messages anyway.

**Comment by aawiseman** [comment]
In addition I think this would help with new mapper retention. People who are actively mapping aren't checking their emails constantly, they are mapping.

**Comment by HolgerJeromin** [comment]
> People who are actively mapping aren't checking their emails constantly, they are mapping.

Most editors shows information about new notification.

**Comment by aawiseman** [comment]
> Most editors shows information about new notification.

iD doesn't, which is where most new mappers come from. I also raised the issue there, someone suggested it was more appropriate for this repo.

**Comment by mxdanger** [comment]
I think you guys are over complicating it design wise. Setting up [Push API](https://w3c.github.io/push-api/) and [Notification API](https://notifications.spec.whatwg.org/#api) is the route we should go. For an overview you can read the following guide: [Introduction to Push Notifications | Web | Google Developers](https://developers.google.com/web/ilt/pwa/introduction-to-push-notifications)

Here's an example I set up.

https://user-images.githubusercontent.com/32040254/116771854-36bdd880-aa02-11eb-9d2d-05777e04b90a.mp4

**Comment by empersOSM** [comment]
Personally, I'd like to see that at least for PMs. If JOSM can do it, osm.org can, too.

**Comment by aawiseman** [comment]
@empersOSM yes, exactly! Let's make it easier to find out that you are getting comments, messages, and so on. Get people to engage with the rest of the community.

**Comment by 1T-money** [comment]
I agree, I think we should be able to get notifications for comments on changesets on the website. It would be better than having to check email.

**Comment by zekefarwell** [comment]
I think something like this is an important feature to add.  It seems quite common for users to not respond at all to changeset comments but then they will respond pretty quickly to private messages.  I'm not entirely sure why, but my theory is that with the private messages they see the unread indicator in the upper right of the website which is also visible while editing in iD.  I don't think a popup or other in-your-face notification would be necessary if changeset discussions showed the same type of unread indicator as messages do.  

I'd imagine having a section very similar to My Messages called My Changeset Discussions.  Instead of Inbox and Outbox it would have a list of your changesets that have comments, and a list of other people's changesets you'd commented on.  In addition, the data model of changeset comments would need to have a read/unread state added so the unread indicator could be shown on the user menu in the upper right.   I imagine this would not be a trivial amount of work, but given how many users don't seem to even be aware that changeset discussions exist, I think it would be very helpful.

**Comment by woodpeck** [comment]
Like TomH said above, this is not a simple "change the data model of changeset comments to add a read/unread state", instead a read/unread state would have to be recorded for everyone who has subscribed to a changeset. -- A less elegant but simpler method would be to generate a user-to-user message to all subscribers every time a changeset comment is created (just like we create emails now). That would give us the read/unread state for free, at the cost of recording duplicates (which users would be free to delete from their inbox).

**Comment by zekefarwell** [comment]
> A less elegant but simpler method would be to generate a user-to-user message to all subscribers every time a changeset comment is created

This seems like a great idea that would get us most of the benefit with less underlying changes required.

**Comment by gravitystorm** [comment]
> > A less elegant but simpler method would be to generate a user-to-user message to all subscribers every time a changeset comment is created
> 
> This seems like a great idea that would get us most of the benefit with less underlying changes required.

From a maintainer's point of view, I have to strongly disagree. What we need (as discussed elsewhere on this tracker) is a  notification system. Different actions trigger different notifications (new user messages, new comments on changesets, closing notes etc). Users can received notifications through various mechanisms (email, on the website, etc). Users can configure which notifications they are interested in, and through what mechanism they want to get each. Something similar to what you see on many other websites, such as here on github, for example.

If we start bodging stuff together (like have "special" user messages, which aren't from a user, and can't be replied to, and aren't in any way part of a conversation between real people) then that just makes it even harder to implement notifications properly, while also ruining the user-to-user messaging system. I'm fine with partial implementations, of course, not everything about notifications needs to be implemented straight away. But I'm absolutely not interested in bodges that make the codebase worse, harder to maintain and harder to build on in future.

**Comment by woodpeck** [comment]
The "as discussed elsewhere" is probably here https://github.com/openstreetmap/openstreetmap-website/issues/1387 though it might have come up in other discussions as well. I hadn't thought of the "specialness" of these messages (aren't from a user, can't be replied to), Andy is of course right here. The reason I had not thought of it is that the emails sent by OSM are just the same happy mix - some are from the system and cannot be replied to (changeset discussions) whereas others originate from user-to-user-messages and can. Repeating this in user-to-user messages would therefore not make things worse, but not better either, and a proper messaging system would of course be preferable.

**Comment by 2dreamy** [comment]
Related discussions in the OSM forum - [About making OSM’s messaging features more efficient](https://community.openstreetmap.org/t/about-making-osms-messaging-features-more-efficient/100050)

**Comment by cmoffroad** [comment]
For immediate improvements, the most readily achievable goal is to enhance the visibility of user badges. I have submitted a proposal in the linked pull request (PR) on GitHub for reference: [GitHub - openstreetmap/openstreetmap-website - PR #4070](https://github.com/openstreetmap/openstreetmap-website/pull/4070#issuecomment-1629242772).

![image](https://github.com/openstreetmap/openstreetmap-website/assets/2485196/65e35449-686e-47fa-9e1d-6a0625d861d8)

**Comment by cmoffroad** [comment]
> From a maintainer's point of view, I have to strongly disagree. What we need (as discussed elsewhere on this tracker) is a notification system. Different actions trigger different notifications (new user messages, new comments on changesets, closing notes etc). Users can received notifications through various mechanisms (email, on the website, etc). Users can configure which notifications they are interested in, and through what mechanism they want to get each. Something similar to what you see on many other websites, such as here on github, for example.

@gravitystorm Since you've also posted a similar comment on a related issue [GitHub - openstreetmap/openstreetmap-website - Issue #4080](https://github.com/openstreetmap/openstreetmap-website/issues/4080), I assume you already have some technical plan in mind. So how do we get started improving the messaging/notifications system? We can break it down into smaller steps and discuss them right away. I am confident there will be plenty of people interested in contributing and making the platform better in that respect. 

As a starter, I would love to see a new "My Changeset Comments (X)" menu item which opens a page similar to NEIS.
The (X) could refer e.g. to unreplied changeset comments on my own changesets. Changeset comments deserve a more prominent and dedicated space!

**Comment by gravitystorm** [comment]
> I assume you already have some technical plan in mind. 

No, I don't have a particular technical plan. I have a rough outline in my mind about how it could work but I haven't spent any time either designing the database and model structure, nor on the user interface. It's just one of many things on the TODO list.

**Comment by kcne** [comment]
Did a bit research on some libraries that offer such functionalities and implementation details that we could potentially use for this:
### 1. [notifications-rails](https://github.com/jonhue/notifications-rails)
`notifications-rails` provides a modular structure for creating, rendering, pushing, and managing settings for notifications, with each module backed by the database. This setup supports on-site notifications with user-configurable preferences, allowing users to see unread counts and mark notifications as read.

**Database Structure and Requirements:**

- **New Tables:**
  - **`notifications`:** Main table storing each notification. Key columns:
    - `target_id` and `target_type` (polymorphic association for the recipient)
    - `object_id` and `object_type` (polymorphic link to the notification’s subject, e.g., a `Comment`)
    - `metadata` (JSON for additional data), `category` (notification type), and `read` (boolean for read/unread).
  - **`notification_groups` (optional):** Allows bulk notifications for groups (e.g., subscribers).

- **Modified Tables:**
  - **User Table**: Adds optional columns for `read_notification_count` and `unread_notification_count` to cache unread counts, and `settings` for storing user-specific notification preferences.

- **Indexes:**  
  - Indexes on `target_id`, `target_type`, and `category` for quick lookups in the `notifications` table.

**Database Workflow:**
1. **Creating Notifications:** Notifications are inserted into `notifications` with details on target, subject, metadata, and category.
2. **Marking as Read:** Updating `read` marks notifications as read, and `unread_notification_count` can be adjusted for caching.
3. **Bulk Notifications:** The `notification_groups` table can manage batch notifications by defining recipient groups.

### 2. [noticed](https://github.com/excid3/noticed)

`noticed` offers individual and bulk notifications, supporting multiple delivery methods, such as ActionCable for real-time updates. It’s suitable for creating on-site notifications tied to specific events, each managed through ActiveJob for asynchronous processing.

**Database Structure and Requirements:**

- **New Tables:**
  - **`notified_events`:** Logs each event that triggers notifications. Key columns:
    - `type` (event type, e.g., “NewCommentNotifier”)
    - `record_id` and `record_type` (polymorphic association to link with the event’s subject, e.g., `Post`)
    - `params` (JSON to store event details, like message and URL).
  
  - **`notifications`:** Links notifications to events and tracks each recipient. Key columns:
    - `event_id` (foreign key to `notified_events`), `recipient_id` and `recipient_type` (polymorphic association for the recipient), and `read_at` (datetime for read status).

- **Modified Tables:**
  - **User Table**: No required modifications, but adding `has_many :notifications` makes it easier to manage notifications.

- **Indexes:**
  - Indexes on `type`, `record_type`, and `record_id` in `notified_events` for efficient filtering and quick retrieval.

**Database Workflow:**
1. **Logging Events:** Each new event (e.g., a comment) is logged in `notified_events`, capturing relevant metadata.
2. **Creating Notifications:** Notifications are linked to events in `notifications`, enabling individual tracking for each recipient.
3. **Read Status:** Users mark notifications as read by updating `read_at`, enabling quick querying for unread notifications and visual badges.

Hope you find this helpful @gravitystorm

**Comment by dp7x** [comment]
Hello guys, I am following this thread with extreme interest. Since two years have passed since last message, are there any news ? I think that a more evident notification for discussed changeset would be a much useful feature to bring users, new and old, to understand the importance of reading and answering the complaints. Maybe, a simple and efficient way could be to trigger an OSM message everytime a new comment is done to all the users involved, exactly how it actually works with emails. This way the user see the number of unread messages when opening ID and is taken to read the notification. Also, a new menu item could be added below "My messages", call it "My comments" or something like that, linking to a page like this one:
https://resultmaps.neis-one.org/osm-discussion-comments?uid=XXXXXXX 
(XXXXXX being the user ID).
The number after "My comments" could be simply the sum of all commented changesets, old and new, without the need to keep count only for unread ones. All these info, as you probably know, can be found also on: https://hdyc.neis-one.org/
